plasmapy/diagnostics/thomson.py

```diff
@@ -152,24 +203,46 @@ def spectral_density(
     .. _`10.5281/zenodo.3766933`: https://doi.org/10.5281/zenodo.3766933
     .. _`Sheffield`: https://doi.org/10.1016/B978-0-12-374877-5.00003-8
     """
+
+    # Validate efract
     if efract is None:
         efract = np.ones(1)
     else:
         efract = np.asarray(efract, dtype=np.float64)
 
+    # Validate ifract
     if ifract is None:
         ifract = np.ones(1)
     else:
         ifract = np.asarray(ifract, dtype=np.float64)
 
-    # If electon velocity is not specified, create an array corresponding
-    # to zero drift
-    if electron_vel is None:
+    # Condition the electron velocity keywords
+    if electron_vel is not None:
+        pass
+    elif (electron_speed is not None) and (electron_vdir is not None):
+        # Normalize vdir: raise exception if all zero
+        electron_vdir = electron_vdir.astype(np.float64)
+        norm = np.linalg.norm(electron_vdir, axis=-1, keepdims=True)
+        if np.any(norm == 0.0):
+            raise ValueError("The electron_vdir vector cannot be zero.")
+        else:
+            electron_vdir *= 1.0 / norm
+        electron_vel = electron_speed[:, np.newaxis] * electron_vdir
```

Comment:
```suggestion
    elif (electron_speed is not None) and (electron_vdir is not None):
        # Normalize vdir: raise exception if all zero
        norm = np.linalg.norm(electron_vdir, axis=-1, keepdims=True)
        if np.any(norm == 0.0):
            raise ValueError("The electron_vdir vector cannot be zero.")
        
        electron_vdir = electron_vdir / norm
        electron_vel = electron_speed[:, np.newaxis] * electron_vdir
```

The `else` is redundant in this case because it's always intended to be reached unless the `Exception` occurs.  You don't have to convert the `dtype` if you don't use the syntax sugar `*=`.  While both option seem to have similar execution times, the one I'm suggestion is a little more stable and (subjectively) I think is a little easier to read.

![image](https://user-images.githubusercontent.com/29869348/112574115-e3c18780-8daa-11eb-9984-d1e9c192f89c.png)